octopus/src/main/scala/octopus/dsl.scala

```diff
@@ -58,118 +58,119 @@ object dsl {
-        (value: U, ec: ExecutionContext) =>
-          v.validate(f(value))(ec)
+    def comap[U](f: U => T): AsyncValidator[M, U] =
+      AsyncValidator.instance[M, U] { (value: U) =>
+        v.validate(f(value))
       }
 
-    def rule(asyncPred: T => Future[Boolean], whenInvalid: String): AsyncValidator[T] =
+    def rule(asyncPred: T => M[Boolean], whenInvalid: String): AsyncValidator[M, T] =
       compose(AsyncValidationRules.rule(asyncPred, whenInvalid))
 
-    def rule[F](selector: T => F, pred: F => Future[Boolean], whenInvalid: String): AsyncValidator[T] =
-      macro DslMacros.ruleFieldSelectorAsync[T, F]
+    def rule[G](selector: T => G, pred: G => M[Boolean], whenInvalid: String): AsyncValidator[M, T] =
+      macro DslMacros.ruleFieldSelectorAsync[M, T, G]
 
-    def ruleVC[V](asyncPred: V => Future[Boolean], whenInvalid: String)
-                 (implicit gen: Generic.Aux[T, V :: HNil]): AsyncValidator[T] =
+    def ruleVC[V](asyncPred: V => M[Boolean], whenInvalid: String)
+                 (implicit gen: Generic.Aux[T, V :: HNil]): AsyncValidator[M, T] =
       compose(AsyncValidationRules.ruleVC(asyncPred, whenInvalid))
 
-    def ruleCatchOnly[E <: Throwable : ClassTag](asyncPred: T => Future[Boolean],
+    def ruleCatchOnly[E <: Throwable : ClassTag](asyncPred: T => M[Boolean],
                                                  whenInvalid: String,
-                                                 whenCaught: E => String): AsyncValidator[T] =
+                                                 whenCaught: E => String): AsyncValidator[M, T] =
       compose(AsyncValidationRules.ruleCatchOnly(asyncPred, whenInvalid, whenCaught))
 
-    def ruleCatchNonFatal(asyncPred: T => Future[Boolean],
+    def ruleCatchNonFatal(asyncPred: T => M[Boolean],
                           whenInvalid: String,
-                          whenCaught: Throwable => String): AsyncValidator[T] =
+                          whenCaught: Throwable => String): AsyncValidator[M, T] =
       compose(AsyncValidationRules.ruleCatchNonFatal(asyncPred, whenInvalid, whenCaught))
 
-    def ruleEither(asyncPred: T => Future[Either[String, Boolean]],
-                   whenInvalid: String): AsyncValidator[T] =
+    def ruleEither(asyncPred: T => M[Either[String, Boolean]],
+                   whenInvalid: String): AsyncValidator[M, T] =
       compose(AsyncValidationRules.ruleEither(asyncPred, whenInvalid))
 
-    def ruleOption(asyncPred: T => Future[Option[Boolean]],
+    def ruleOption(asyncPred: T => M[Option[Boolean]],
                    whenInvalid: String,
-                   whenNone: String): AsyncValidator[T] =
+                   whenNone: String): AsyncValidator[M, T] =
       compose(AsyncValidationRules.ruleOption(asyncPred, whenInvalid, whenNone))
 
   }
 
-  implicit class AsyncValidatorSyncOps[T](val v: AsyncValidator[T]) extends AnyVal {
+  implicit class AsyncValidatorSyncOps[M[_]: App, T](val v: AsyncValidator[M, T]) {
```

Comment:
The same as above: dropped `extends AnyVal`.

Reply:
`extends AnyVal` is back there